wave_vision/src/matcher/brute_force_matcher.cpp

```diff
@@ -53,28 +65,133 @@ void BruteForceMatcher::checkConfiguration(
+    // distance
+    for (auto &match : matches) {
+        if (match.distance <=
+            this->current_config.distance_threshold * min_distance) {
+            filt_matches.push_back(match);
+        }
+    }
+
+    return filt_matches;
+}
+
+std::vector<cv::DMatch> BruteForceMatcher::filterMatches(
+  std::vector<std::vector<cv::DMatch>> &matches) const {
+    std::vector<cv::DMatch> filt_matches;
+    float ratio;
+
+    for (auto &match : matches) {
+        // Calculate ratio between two best matches. Accept if less than
+        // ratio heuristic
+        ratio = match[0].distance / match[1].distance;
+        if (ratio <= this->current_config.ratio_threshold) {
+            filt_matches.push_back(match[0]);
+        }
+    }
+
+    return filt_matches;
+}
+
+std::vector<cv::DMatch> BruteForceMatcher::removeOutliers(
+  const std::vector<cv::DMatch> &matches,
+  const std::vector<cv::KeyPoint> &keypoints_1,
+  const std::vector<cv::KeyPoint> &keypoints_2) const {
+    std::vector<cv::DMatch> good_matches;
+    std::vector<cv::Point2f> fp1, fp2;
+
+    // Take all good keypoints from matches, convert to cv::Point2f
+    for (auto &match : matches) {
+        fp1.push_back(keypoints_1.at(match.queryIdx).pt);
+        fp2.push_back(keypoints_2.at(match.trainIdx).pt);
+    }
+
+    // Find fundamental matrix
+    std::vector<uchar> mask;
+    cv::Mat fundamental_matrix;
+
+    fundamental_matrix =
+      cv::findFundamentalMat(fp1, fp2, cv::FM_RANSAC, 3., 0.99, mask);
+
+    // Only retain the inliers matches
+    for (size_t i = 0; i < mask.size(); i++) {
+        if (mask.at(i) != 0) {
+            good_matches.push_back(matches.at(i));
+        }
+    }
+
+    return good_matches;
 }
 
 std::vector<cv::DMatch> BruteForceMatcher::matchDescriptors(
-  const cv::Mat &descriptors_1, const cv::Mat &descriptors_2) const {
-    std::vector<cv::DMatch> matches;
-
-    /** Mask variable, currently unused.
-     *
-     *  The mask variable indicates which descriptors can be matched between the
-     *  two sets. As per OpenCV docs "queryDescriptors[i] can be matched with
-     *  trainDescriptors[j] only if masks.at<uchar>(i,j) is non-zero.
-     *
-     *  In the libwave wrapper, queryDescriptors and trainDescriptors are
-     *  referred to as descriptors_1 and descriptors_2.
-     */
-    cv::InputOutputArray mask = cv::noArray();
-
-    // Determine matches between sets of descriptors
-    this->brute_force_matcher->match(
-      descriptors_1, descriptors_2, matches, mask);
-
-    return matches;
+  const cv::Mat &descriptors_1,
+  const cv::Mat &descriptors_2,
+  const std::vector<cv::KeyPoint> &keypoints_1,
+  const std::vector<cv::KeyPoint> &keypoints_2,
+  const cv::InputArray &mask) const {
+    std::vector<cv::DMatch> good_matches;
+    std::vector<cv::DMatch> filtered_matches;
+
+    if (current_config.use_knn) {
+        std::vector<std::vector<cv::DMatch>> matches;
+        int k = 2;
```

Comment:
okay after discussion with @Jebediah, the decision is to include them in the BFMatcherParams struct, but not include it as a variable in the constructor. Therefore, the user cannot change this.